ztp/gitops-subscriptions/argocd/SNOExpansion.md

```diff
@@ -0,0 +1,483 @@
+{"lastTransitionTime":"2022-08-21T08:39:04Z","message":"Resource action is done.","reason":"ActionDone","status":"True","type":"Completed"}
+```
+
+Delete the managedclusterview to free up resources:
+
+```bash
+$ oc delete managedclusterview check-sriov-node-selector -n $SNO_NAME
+```
+##### __Redeploy SNO policies__
+Redeploy the common policies using ztp plugin container version 4.12 or above using TALM operator custom resource.
+
+#### __Preparing worker node policies__
+Create the following policy template:
+
+```yaml
+apiVersion: ran.openshift.io/v1
+kind: PolicyGenTemplate
+metadata:
+  name: "cnfdf15-worker"
+  namespace: "ztp-cnfdf15-policies"
+spec:
+  bindingRules:
+    # These policies will correspond to all clusters with this label:
+    sites: "cnfdf15"
+######
+# Comment 1
+  mcp: "worker"
+  sourceFiles:
+    - fileName: PtpConfigSlave.yaml   # Change to PtpConfigSlaveCvl.yaml for ColumbiaVille NIC
+      policyName: "config-policy"
+      metadata:
+        name: "du-ptp-slave-worker"
+      spec:
+        profile:
+        - name: "slave-worker"
+          # This interface must match the hardware in this group
+######
+# Comment 2
+          interface: "ens1f0"
+          ptp4lOpts: "-2 -s --summary_interval -4"
+          phc2sysOpts: "-a -r -n 24"
+        recommend:
+        - match:
+          - nodeLabel: kubernetes.io/hostname=cnfdf16
+          priority: 4
+          profile: slave-worker
+    - fileName: SriovNetwork.yaml
+######
+# Comment 3    
+      policyName: "config-policy-net"
+      metadata:
+        name: "f1u-network-worker"
+      spec:
+        resourceName: uplane_worker
+        vlan: 140
+    - fileName: SriovNetworkNodePolicy.yaml
+      policyName: "config-policy-net"
+      metadata:
+        name: "sriov-nnp-uplane-worker"
+      spec:
+        deviceType: vfio-pci
+        isRdma: false
+        nicSelector:
+          pfNames: ["ens1f1"]
+        nodeSelector:
+          kubernetes.io/hostname: cnfdf16
+        numVfs: 8
+        priority: 10
+        resourceName: uplane_worker    
+    
+    - fileName: SriovNetwork.yaml
+      policyName: "config-policy-net"
+      metadata:
+        name: "f1c-network-worker"
+      spec:
+        resourceName: cplane_worker
+        vlan: 150
+
+    - fileName: SriovNetworkNodePolicy.yaml
+      policyName: "config-policy-net"
+      metadata:
+        name: sriov-nnp-cplane-worker
+      spec:
+        deviceType: netdevice
+        isRdma: true
+        nicSelector:
+          pfNames: ["ens1f0"]
+        nodeSelector:
+          kubernetes.io/hostname: cnfdf16
+        numVfs: 8
+        priority: 10
+        resourceName: cplane_worker
+######
+# Comment 4  
+    - fileName: MachineConfigGeneric.yaml
+      policyName: "config-policy"
+      metadata:
+        labels:
+          machineconfiguration.openshift.io/role: worker
+        name: enable-workload-partitioning
```

Comment:
It may be worth having a section to discuss how workload partitioning works with SNO expansion. Checking my understanding here.... 
The expansion node will be a worker, so only those core openshift pods that are not bound (nodeSelector) to the master will run on this node. For pods which are not bound to the master node I suspect the scheduler won't place them on this expansion node until this workload partitioning config is put in place (since there are no `management.workload.openshift.io/cores` resources advertised).
Is that accurate?

Reply:
Not sure then how any of the daemonset pods would be running. But they are... Let me check this

Reply:
Unless they are daemonset pods. It seems to be enough to apply the machineconfig defining the wl partitioning and reboot the node. 
I will have a deeper look into this. And for sure it should be documented somewhere. Is here the right place for it?

Reply:
I think I got it. Without the workload partitioning configured on the kubelet side, the pods would not be mutated to the new resource type. The resource they request will remain "cpu", and not management.workload.openshift.io/cores.
For example, this is the worker without WL partitioning enabled:
```bash
oc -n openshift-machine-config-operator describe po/machine-config-daemon-lp2js
...
    Requests:
      cpu:     20m
      memory:  50Mi
```
This is its "brother" on the master with WL partitioning enabled:
```bash
oc -n openshift-machine-config-operator describe po/machine-config-daemon-rsw44
...
Requests:
      management.workload.openshift.io/cores:  20
```
How the API server knows which pods to mutate and which not to? I guess it checks the kubelet capabilities  when the node boots and registers with the API.

Reply:
The admission hook, checks for the following: 
1) Checks that it is a SNO
2) Checks that all nodes in the cluster are advertising the management.workload.openshift.io/cores
3) namespace has the following annotation: workload.openshift.io/allowed: management
4) pod has the following annotation: target.workload.openshift.io/management: {"effect":"PreferredDuringScheduling"}